OpenSimRT/RealTime/experimental/RealTimeAnalysisExtended/tests/TestRTExtFromFile.cpp

```diff
@@ -0,0 +1,320 @@
+/**
+ * -----------------------------------------------------------------------------
+ * Copyright 2019-2021 OpenSimRT developers.
+ *
+ * This file is part of OpenSimRT.
+ *
+ * OpenSimRT is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * OpenSimRT is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+ * A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * OpenSimRT. If not, see <https://www.gnu.org/licenses/>.
+ * -----------------------------------------------------------------------------
+ *
+ * @file TestRTExtFromFile.cpp
+ *
+ * @brief Tests the RealTimeAnalysisExtended class with data acquired from file.
```

Comment:
It is good to comment on the delay that you observed in your tests. We might need to run this on a faster computer because of the GRF prediction that might delay the pipeline.

Reply:
The GRF prediction is quite fast. I've measured the performance of the pipelines, and there is no significant difference in delay(FPS) between to original and extended pipeline. However, the comparison is not very reliable as thread priorities and system load affects heavily the measurements. Both pipelines result in ~30ms delay (without SO+JR).

Reply:
I think 30ms is a lot without SO. What is causing this delay? How come GRF prediction/not does not affect the pipeline? Do the unit tests indicate a delay of 10ms or is it due to the streamer?

Reply:
The total processing in the acquisition thread lasts ~13ms. This is added on-top of the delay in the `dataAcquisitionFunction` passed to the pipeline which is 16ms. Resulting in 29ms delay even without the SO.

Reply:
I used to have a separate `streaming` thread for the streaming data in a constant rate, and the `dataAcquisitionFunction` would only fetch data from that stream, without the added delay. Back then, however, it was problematic to open an additional thread for streaming. I'll look back into it now that I have changed OS.

Reply:
I remember that the pipeline was very efficient in v3.3, even with SO. We have to look at how to redesign it so that everything is efficient.

Reply:
![image](https://user-images.githubusercontent.com/26690578/111364697-3e1b6380-869a-11eb-835e-bd0069c4bf53.png)

Reply:
All the processing goes to filtering (due to splines) and SO. I don't know if those can be improved any further.

Reply:
![image](https://user-images.githubusercontent.com/26690578/111365640-6788bf00-869b-11eb-9c31-92d7133bca3a.png)

And here's another without the SO.

Reply:
![1615946083769](https://user-images.githubusercontent.com/4024976/111402878-47bdbf00-86cc-11eb-9236-c3d75455dc05.jpg)

I have put worst-case numbers. In this case, the acquisition should have an upper limit of 140 FPS and the processing thread 45 FPS. Ensure that data comes to the acquisition thread at frames lower than 140 (e.g., 100 is a good choice). You only have to process the last frame in the processing thread since we do not have any state.

If we move the filter to the processing thread, then the acquisition thread can achieve high FPS (e.g., >500), but the processing will be around 30 FPS. It is best if we can keep the filter on the acquisition thread.

Reply:
Then, we also have to minimize the read/write time to the buffer so that filtered data can be easily mapped to and from.

Reply:
> All the processing goes to filtering (due to splines) and SO. I don't know if those can be improved any further.

This is a good point. Maybe in the future, we can improve the spline and filtering to reuse data from the previous step. For now, let us assume that the filter operates at 6ms worst-case and we may try to improve this.

Reply:
That's exactly the pipeline followed right now. But you forget the `dataAcquisitionFunction` is a function that lasts 16ms, which are added to the acquistion time, resulting in max 43fps. In my system I don't have the same measured delays as shown in the picture, so I get around 30ms instead.

Reply:
OK, I see, but you should not account for this delay. The acquisition thread should execute quickly if you do not add this delay. What is then the execution time of the processing thread with SO? This will dictate the throughput of the system.

Reply:
Around 43ms.

Reply:
Why 43? Can you specify the delay of each operation? I would expect SO to be around 20. The rest should execute fast.